File: cracking_the_coding_interview/10-10.cpp
#include <iostream>
#include <unordered_map>
using namespace std;

class RankOfNumber {
public:
    RankOfNumber() : root(nullptr){}
    void track(int x){
        root = track(x, root);
    }

    int getRankOfNumber(int x){
        return getRankOfNumber(x, 0, root);
    }

private:
    unordered_map<int, int> rankMap;
    struct TreeNode {
        int val;
        int count;
        TreeNode* left;
        TreeNode* right;
    };
    TreeNode* root;
    TreeNode* track(int x, TreeNode* node) {
        if(!node){
            node = new TreeNode();
            node->val = x;
            node->count = 0;
            node->left = nullptr;
            node->right = nullptr;
            return node;
        }
        // cout << node->val << endl;
        if(x == node->val){
            node->count++;
            return node;
        }
        if(x < node->val){
            node->count++;
            TreeNode* next = track(x, node->left);
            if(!(node->left))
                node->left = next;
            return node;
        }
        TreeNode* next = track(x, node->right);
        if(!(node->right))
            node->right = next;
        return node;
    }

    int getRankOfNumber(int x, int count, TreeNode* node){
        // cout << node->val << endl;
        if(x == node->val)
            return count + node->count;
        if(x < node->val)
            return getRankOfNumber(x, count, node->left);
        return getRankOfNumber(x, count + node->val + 1, node->right);
    };
};

int main(void){
    RankOfNumber ron;
    ron.track(5);
    ron.track(1);
    ron.track(4);
    ron.track(4);
    ron.track(5);
    ron.track(9);
    ron.track(7);
    ron.track(13);
    ron.track(3);
    cout << ron.getRankOfNumber(1) << endl; // 0
    cout << ron.getRankOfNumber(3) << endl; // 1
    cout << ron.getRankOfNumber(4) << endl; // 3
    return 0;
}

File: cracking_the_coding_interview/10-10.py
import unittest

class TreeNode:
    def __init__(self, val, count=0, left=None, right=None):
        self.val = val
        self.count = count
        self.left = left
        self.right = right

class RankOfNumber:
    def __init__(self):
        self.root = None
    
    def trace(self, x):
        self.root = self.__trace(x, self.root, 0)
    
    def get_rank_of_number(self, x):
        return self.__get_rank_of_number(x, self.root)
    
    def __trace(self, x, node, count):
        if not node:
            node = TreeNode(x, count)
            return node
        
        if x == node.val:
            node.count += 1
        elif x < node.val:
            node.count += 1
            node.left = self.__trace(x, node.left, count)
        elif x > node.val:
            node.right = self.__trace(x, node.right, count + 1)
        return node
    
    def __get_rank_of_number(self, x, node):
        if not node:
            return 0
        if node.val == x:
            return node.count
        if node.val > x:
            return self.__get_rank_of_number(x, node.left)
        if node.val < x:
            return self.__get_rank_of_number(x, node.right) + node.count

class Test(unittest.TestCase):
    def test_simple(self):
        ron = RankOfNumber()
        ron.trace(5)
        ron.trace(1)
        ron.trace(4)
        ron.trace(4)
        ron.trace(5)
        ron.trace(9)
        ron.trace(7)
        ron.trace(13)
        ron.trace(3)
        self.assertEqual(ron.get_rank_of_number(1), 0)
        self.assertEqual(ron.get_rank_of_number(3), 1)
        self.assertEqual(ron.get_rank_of_number(4), 3)

if __name__ == '__main__':
    unittest.main()
